src/ledger/LedgerTxn.cpp

```diff
@@ -57,6 +58,156 @@ AbstractLedgerTxnParent::~AbstractLedgerTxnParent()
 {
 }
 
+// Implementation of AbstractPrefetcher -------------------------------------
+AbstractPrefetcher::~AbstractPrefetcher()
+{
+}
+
+// Implementation of Prefetcher ---------------------------------------------
+LedgerTxnRoot::Impl::Prefetcher::Prefetcher(size_t batchSize)
+    : mBatchSize(batchSize)
+{
+}
+
+LedgerTxnRoot::Impl::Prefetcher::~Prefetcher()
+{
+}
+
+std::list<LedgerKey const>&
+LedgerTxnRoot::Impl::Prefetcher::getQueueByType(LedgerEntryType let)
+{
+    switch (let)
+    {
+    case ACCOUNT:
+        return mAccountsQueue;
+    case DATA:
+        return mDataQueue;
+    case OFFER:
+        return mOffersQueue;
+    case TRUSTLINE:
+        return mTrustlinesQueue;
+    default:
+        throw std::runtime_error("Unknown ledger entry type");
+    }
+}
+
+void
+LedgerTxnRoot::Impl::Prefetcher::queueForPrefetch(LedgerKey const& key,
+                                                  bool immediate)
+{
+    auto& queue = getQueueByType(key.type());
+    if (immediate)
+    {
+        queue.push_front(key);
+    }
+    else
+    {
+        queue.push_back(key);
+    }
+}
+
+void
+LedgerTxnRoot::Impl::Prefetcher::cancelPrefetch(LedgerKey const& key)
+{
+    auto& queue = getQueueByType(key.type());
+    auto toCancel = std::find(queue.begin(), queue.end(), key);
+    if (toCancel != queue.end())
+    {
+        queue.erase(toCancel);
+    }
+}
+
+void
+LedgerTxnRoot::Impl::Prefetcher::cancelAllPrefetches()
+{
+    mAccountsQueue.clear();
+    mDataQueue.clear();
+    mOffersQueue.clear();
+    mTrustlinesQueue.clear();
+}
+
+bool
+LedgerTxnRoot::Impl::Prefetcher::hasNextToPrefetch(LedgerEntryType type)
+{
+    auto const& queue = getQueueByType(type);
```

Comment:
then calling from `getBatchToPrefetch`, this is calling `getQueueByType` mBatchSize times even when result will always be the same